### PR TITLE
[mtp] add rejection sampling for speculative decoding

### PR DESCRIPTION
**Pull request by liyucheng09**
## Motivation

Add classic reject sampling (RS) in Eagle speculative decoding.

The current pipeline of Eagle speculative decoding employ greedy sampling in both predicting draft tokens and target verification.

Use classic reject sampling means to use actual draft prob distribution for sampling draft tokens and for acceptance criterion.

Experimental results show using classic rs can largely improve accept length across various tasks.

## Modifications

We change how tokens are drafted: from argmax to sample on draft distribution. And we provide a reject sample triton kernel to update accept criterion from `p` to `min(1, p\q)`.

## Accuracy

unit test results with `Qwen/Qwen3.5-9B`:

```
[METRIC] gsm8k_score=0.905 labels={"model": "/cpfs01/user/lyc508653/sync/checkpoints/hf/qwen35-9b", "eval": "gsm8k"}
[METRIC] gsm8k_latency=16.090488217771053 labels={"model": "/cpfs01/user/lyc508653/sync/checkpoints/hf/qwen35-9b", "eval": "gsm8k"}
Writing report to /tmp/gsm8k__cpfs01_user_lyc508653_sync_checkpoints_hf_qwen35-9b.html
{'score:std': np.float64(0.293214938227915), 'score': np.float64(0.905), 'latency': 16.090488217771053, 'output_throughput': 1404.6186600502554}
Writing results to /tmp/gsm8k__cpfs01_user_lyc508653_sync_checkpoints_hf_qwen35-9b.json
metrics={'score:std': np.float64(0.293214938227915), 'score': np.float64(0.905), 'latency': 16.090488217771053, 'output_throughput': 1404.6186600502554}
```

## Speed Tests and Profiling

- model=Qwen/Qwen3.5-397B-A17B-FP8
- tp=8, dp=4, bsz=64, outputlen=16k

- Dataset: `atcoder python`
- temperature=1

| Metric | No MTP | MTP | MTP+RS |
|--------|-------:|----:|-------:|
| Throughput (tok/s) | 1,640.34 | 3,280.80 (+100.0%) | 3,785.60 (+130.8%) |
| Accept Rate | — | 0.6876 | 0.7896 (+14.8%) |
| Accept Length | — | 3.06 | 3.37 (+10.0%) |
| Verify Count | — | 333,849 | 300,011 (-10.1%) |
| Step-1 Accept Ratio | — | 0.8231 | 0.9150 (+11.2%) |
| Step-2 Accept Ratio | — | 0.6759 | 0.7899 (+16.9%) |
| Step-3 Accept Ratio | — | 0.5562 | 0.6599 (+18.6%) |

- temperature=0.6

| Metric | No MTP | MTP | MTP+RS |
|--------|-------:|----:|-------:|
| Throughput (tok/s) | 1,650.60 | 3,496.49 (+111.8%) | 3,662.92 (+121.9%) |
| Accept Rate | — | 0.7773 | 0.8191 (+5.4%) |
| Accept Length | — | 3.33 | 3.46 (+3.8%) |
| Verify Count | — | 301,375 | 297,365 (-1.3%) |
| Step-1 Accept Ratio | — | 0.8902 | 0.9261 (+4.0%) |
| Step-2 Accept Ratio | — | 0.7740 | 0.8196 (+5.9%) |
| Step-3 Accept Ratio | — | 0.6645 | 0.7080 (+6.5%) |

- Dataset: `deepmath`
- temperature=1

| Metric | No MTP | MTP | MTP+RS |
|--------|-------:|----:|-------:|
| Throughput (tok/s) | 2,365.02 | 4,587.43 (+94.0%) | 4,929.50 (+108.4%) |
| Accept Rate | — | 0.7341 | 0.8185 (+11.5%) |
| Accept Length | — | 3.20 | 3.46 (+7.9%) |
| Verify Count | — | 314,038 | 288,101 (-8.3%) |
| Step-1 Accept Ratio | — | 0.8527 | 0.9290 (+9.0%) |
| Step-2 Accept Ratio | — | 0.7238 | 0.8219 (+13.6%) |
| Step-3 Accept Ratio | — | 0.6149 | 0.7041 (+14.5%) |

- temperature=0.6

| Metric | No MTP | MTP | MTP+RS |
|--------|-------:|----:|-------:|
| Throughput (tok/s) | 2,334.44 | 4,781.54 (+104.8%) | 4,916.06 (+110.6%) |
| Accept Rate | — | 0.8105 | 0.8333 (+2.8%) |
| Accept Length | — | 3.43 | 3.50 (+2.0%) |
| Verify Count | — | 277,181 | 270,616 (-2.4%) |
| Step-1 Accept Ratio | — | 0.9114 | 0.9342 (+2.5%) |
| Step-2 Accept Ratio | — | 0.8091 | 0.8356 (+3.3%) |
| Step-3 Accept Ratio | — | 0.7077 | 0.7298 (+3.1%) |

- model: `zai-org/GLM-5`
- dataset: `deepmath`
- temperature=1

| Metric | No MTP | MTP | MTP+RS |
|--------|-------:|----:|-------:|
| Throughput (tok/s) | 1,226.78 | 2,538.90 (+107.0%) | 2,642.54 (+115.4%) |
| Accept Rate | — | 0.6855 | 0.7501 (+9.4%) |
| Accept Length | — | 3.06 | 3.25 (+6.3%) |
| Verify Count | — | 286,352 | 260,021 (-9.2%) |
| Step-1 Accept Ratio | — | 0.8281 | 0.9071 (+9.5%) |
| Step-2 Accept Ratio | — | 0.6748 | 0.7549 (+11.9%) |
| Step-3 Accept Ratio | — | 0.5372 | 0.5829 (+8.5%) |

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

coauthors: @iofu728 @yizhang2077 @cao1zhg 

---
### CI States

Latest PR Test (Base): :x: [Run #27882600655](https://github.com/sgl-project/sglang/actions/runs/27882600655)
Latest PR Test (Extra): :x: [Run #27882600604](https://github.com/sgl-project/sglang/actions/runs/27882600604)

### CODE REVIEW

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py [hunk 1]
```diff
@@ -140,6 +147,11 @@ def __init__(
                 else None
             )
 
+            self.temperatures = torch.ones((self.max_bs, 1), dtype=torch.float)
+            self.top_ps = torch.ones((self.max_bs,), dtype=torch.float)
```
Comment (Qiaolin-Yu) [inline python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: qq: where are these buffer used?

**Comment by Qiaolin-Yu** [comment]
could you fix the conflicts? ty!

File: python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py (hunk 1 above)
Comment (liyucheng09) [reply python/sglang/srt/speculative/eagle_draft_cuda_graph_runner.py]: hi qianlin, temperature is used in draft prob renorm at: https://github.com/sgl-project/sglang/pull/26312/changes#diff-842a36702d0063dc2d3f6f14ef88dae596fdc2216ba6b4f1ef6c00e2513c82e8R1325

And topp/topk are redundant and just being removed in the lastest commit.

**Comment by liyucheng09** [comment]
> could you fix the conflicts? ty!

@Qiaolin-Yu I just merge from the main, and it should be conflict-free now. thx!

**Comment by haiduo** [comment]
Speed Tests and profiling are Qwen/Qwen3.5-9B or Qwen/Qwen3.5-397B-A17B-FP8?

**Review by Qiaolin-Yu — APPROVED** [review]

**Comment by Qiaolin-Yu** [comment]
/rerun-test test/registered/spec/eagle/test_eagle_reject_sampling.py

**Comment by Qiaolin-Yu** [comment]
/rerun-test test/registered/spec/eagle/test_eagle_reject_sampling.py

**Comment by Qiaolin-Yu** [comment]
/rerun-test test/registered/spec/eagle/test_eagle_reject_sampling.py

**Comment by Qiaolin-Yu** [comment]
/rerun-test test/registered/spec/eagle/test_eagle_reject_sampling.py

**Comment by dongyibo** [comment]
@liyucheng09  Hello, this feature seems to be incompatible with PD disagg

**Comment by DAntyNoel** [comment]
Hi authors, I recently read your excellent paper, **Breaking Entropy Bounds: Accelerating RL Training via MTP with Rejection Sampling**, with great interest. I found it highly insightful. It not only systematically analyzes the relationship between changes in model entropy and the decline in MTP acceptance rate during RL, but also presents clear and practical solutions based on rejection sampling and e2e TV loss.

Coincidentally, I have also been working recently on MTP, RL post-training, and long-sequence rollout acceleration. Your paper has provided an important reference for my research and sparked my interest in pre-RL MTP SFT, rejection sampling, and strategies for updating MTP heads during RL.

After reading the paper and the related pull requests, my understanding is that the MTP heads of Qwen3.5-35A3B are trained on “mixed RFT data” during the pre-RL/SFT stage. However, while trying to understand and follow up on this work more closely, I found that some details regarding the data and training pipeline do not appear to be included in the paper or the currently available public code. To avoid misinterpreting your experimental setup through my own assumptions, I would be grateful if you could clarify the following questions.

RFT data format and chain of thought

- What exactly does “RFT” refer to in the paper?
- Does each MTP training example contain a single-turn completion or a complete trajectory?
- Are explicit chain-of-thought (CoT) or reasoning tokens retained and included in the MTP loss?
- For Agent/SWE data, how are tool calls and tool responses represented, and which fields or tokens are masked out?

Data mixture

- Which data domains were included in the primary and extended experiments, such as code, agent/SWE, reasoning, instruction/general, and others?
- Would it be possible to provide the post-filtering sample counts, token counts, or sampling weights for each domain?

RFT dataset size

- How many raw candidate samples, retained trajectories, packed sequences, effective tokens, and optimizer steps were used?
- How was “one epoch” defined in the paper?
- Does 256K refer to a fixed packed-sequence length or a maximum sequence length?

Use of RFT data during RL

- Was the offline RFT data used only to produce the pre-RL MTP checkpoint, or was it also read or replayed during RL?
- For the online MTP update experiments in Section 6.3, was the MTP loss computed using trajectories generated by the current RL rollout policy, separate offline RFT batches, or a mixture of both?

Relationship between the SFT/RFT and RL training task sets

- Were the prompt/task sets used for RL training identical to, a filtered subset of, partially overlapping with, or completely separate from the pre-RL RFT set?
- What data splitting, filtering, deduplication, and evaluation-set contamination checks were applied?

If the original training data cannot be shared, an anonymized data manifest or a table containing only aggregate statistics would already be sufficient. We would also be very grateful for answers to any subset of these questions that could help us move the reproduction effort forward.

Thank you very much for your time.
